src/state/opamFormatUpgrade.ml

```diff
@@ -1347,38 +1351,96 @@ let as_necessary ?reinit requested_lock global_lock root config =
-  | _, _ -> ()
+  if not changes then None else
+    let root = gt.root in
+    let config = gt.config in
+    let config_f = OpamPath.config gt.root in
+    (* If we don't have a written opam root version in a config file,
+       we are unable to determine if there is an upgrade to do. This can happen in
+       case there is only on the fly upgrades from 2.0. Should we fail ? *)
+    let written_root_version = OpamFile.Config.raw_root_version config_f
+    in
+    (* Config already upgraded *)
+    if OpamStd.Option.equal OpamVersion.equal
+        written_root_version
+        (Some OpamFile.Config.root_version) then None else
+      let written_root_version =
+        OpamStd.Option.default v2_0 written_root_version
+      in
+      let updates =
+        List.filter (fun (v,_) ->
+            OpamVersion.compare written_root_version v < 0)
+          updates
+      in
+      match lock_kind with
+      | `Lock_none | `Lock_read ->
+        (* apply repo or state config updates *)
+        List.fold_left (fun rs_config (_v,from) ->
+            from ?config:rs_config root config)
+          None updates
+      | `Lock_write ->
+        (* If a write lock is required, we need to run through the upgrade
+           mechanism from the beginning to enforce a write at each step if
+           needed *)
+        let is_dev = OpamVersion.is_dev_version () in
+        OpamConsole.errmsg "%s" @@
+        OpamStd.Format.reformat @@
+        Printf.sprintf
+          "This %sversion of opam requires an update to the layout of %s \
+           from version %s to version %s, which can't be reverted.\n\
+           You may want to back it up before going further.\n"
+          (if is_dev then "development " else "")
+          (OpamFilename.Dir.to_string gt.root)
+          (OpamVersion.to_string written_root_version)
+          (OpamVersion.to_string (OpamFile.Config.opam_root_version config));
+        if OpamConsole.confirm "Continue?" then
+          flock_root `Lock_write root @@ fun _ ->
+          (* we keep only light upgrades as hard upgrade is already handled by
+             global state loading, so we must not have to handle hard upgrades
+             as this point. *)
+          let _, upgrades = upgrades written_root_version in
+          let config =
+            OpamStd.Option.default config
+              (OpamFile.Config.BestEffort.read_opt config_f)
+          in
+          let config =
+            List.fold_left (fun config (v, from) ->
+                let config, _change = from ~on_the_fly:false root config in
+                config |> OpamFile.Config.with_opam_root_version v)
+              config upgrades
+          in
+          OpamFile.Config.write (OpamPath.config root) config;
+          erase_plugin_links root;
+          read_f root
+        else
+          OpamStd.Sys.exit_because `Aborted
+
+let as_necessary_repo lock_kind gt =
+  let updates = [
+  ] in
+  as_necessary_repo_switch_t
```

Comment:
```suggestion
let as_necessary_repo lock_kind gt =
  if gt.global_state_to_upgrade.gtc_repo then None else
  let updates = [
  ] in
  as_necessary_repo_switch_t
```

Implem details: I'm not if its better, there is pro & cons:
* as is: factorise the "no change" case like that we ensur same beahviour between repo & switch
* suggestion: no need to define all these arguments if nothing is done, which is 99% of cases.